### PR TITLE
#4: loading decorator should be attached to the UI component (closes #4)

### LINKED ISSUE #4: loading decorator should be attached to the UI component
## requirements
- it should work with https://github.com/buildo/react-avenger/pull/13/files
- as less breaking as possible (but still, breaking)

## specs
- no loadingDecorator wrapping the container component
- as it is an incremental change, leave it in the configuration, but make it wrap the ui component
- optional render logic is handled in container's getLocals

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by francescogior**
Issue #4

### tests performed
- merge and `npm run release` 
  - https://github.com/buildo/react-avenger/pull/13
- loading decorator doesn't wrap the container wrapper anymore
- (it is temporarily used to wrap the `Component` (through `skinnable` and `contains`) for compatibility in order to get an incremental backporting in the projects)
- the container always pass down `readyState` (if not overridden from `mapProps`) in order to let the wrapped `Component` to use normally the `loading` decorator
- the optional rendering logic performed before by `loading` is now handled by a `isReady` function (`props: any => Boolean`, with default the `defaultIsReady` from `react-avenger/loading`)  and `getLocals`

- [x] **BEFORE MERGE, update react-avenger version in package.json**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,16 +1,28 @@
...
-import noLoaderLoading from './noLoaderLoading';
 import displayName from './displayName';
+import noLoaderLoading from './noLoaderLoading';
+
+const isFetched = ({ readyState, ...props }) => {
+  return every(readyState, (rs, k) => (
+    props[k] !== void 0 && typeof rs.error === 'undefined'
+  ));
+};
+
```
Comment (francescogior) [inline src/index.js]: this is basically the old default `isReady` from `react-avenger/loading`

File: src/index.js [hunk 2]
```diff
@@ -1,16 +1,28 @@
...
+const stripUndef = omitF(isUndefined);
 
 const ContainerConfig = t.interface({
+  waitForQueryProps: t.maybe(t.Boolean),
```
Comment (francescogior) [inline src/index.js]: this should be replaced with a `isReady` function.

File: src/index.js [hunk 3]
```diff
@@ -1,16 +1,21 @@
...
 import { t, props } from 'tcomb-react';
 import { skinnable, pure, contains } from 'revenge';
 import _declareConnect from 'buildo-state/lib/connect';
 import _declareQueries from 'react-avenger/lib/queries';
 import _declareCommands from 'react-avenger/lib/commands';
-import noLoaderLoading from './noLoaderLoading';
+import { defaultIsReady, noLoaderLoading } from 'react-avenger/lib/loading';
 import displayName from './displayName';
 
+const stripUndef = omitF(isUndefined);
```
Comment (FrancescoCioria) [inline src/index.js]: rename to `omitUndef`? :)

File: src/index.js [hunk 4]
```diff
@@ -33,10 +38,12 @@ const PublicDecoratorConfig = DecoratorConfig.extend({
...
 const reduceQueryPropsReturn = queries => t.interface({
   accumulator: t.Any,
-  props: t.interface(
-    queries.reduce((ac, k) => ({ ...ac, [k]: t.Any }), {}),
-    { strict: true }
-  )
+  props: t.interface({
+    ...queries.reduce((ac, k) => ({ ...ac, [k]: t.Any }), {}),
+    readyState: t.interface(queries.reduce((ac, k) => ({ ...ac, [k]: t.interface({
+      waiting: t.Boolean, fetching: t.Boolean, loading: t.Boolean, error: t.maybe(t.Any), ready: t.maybe(t.Boolean)
```
Comment (FrancescoCioria) [inline src/index.js]: why is `ready` optional?

File: src/index.js [hunk 5]
```diff
@@ -109,12 +115,22 @@ const decorator = ({ declareQueries, declareCommands, declareConnect }) => (Comp
...
+
     static displayName = displayName(Component, 'Container');
-    getLocals = getLocals;
+
+    getLocals(props) {
+      const { readyState } = props;
+      if (isReady(props)) {
+        return { ...getLocals(props), ...stripUndef({ readyState }) };
+      } else {
+        return stripUndef({ readyState });
```
Comment (FrancescoCioria) [inline src/index.js:134]: **minor and subjective** I prefer APIs that return a JSON with always the same keys: I would not strip the `undefined`s (which IMO should only be `error`) and instead set it to `null`

File: src/index.js (hunk 4 above)
Comment (FrancescoCioria) [reply src/index.js]: you've missed a `{ strict: true }` here.
To simplify this code I would declare outside this block a
```js
const ReadyState = t.interface({
  waiting: t.Boolean,
  fetching: t.Boolean,
  loading: t.Boolean,
  error: t.maybe(t.Any),
  ready: t.maybe(t.Boolean) // why optional??
}, { strict: true });
```

Comment (francescogior) [reply src/index.js]: 1. Probably for caution, it should be safe to remove the `maybe`
2. good idea 👍

File: src/index.js (hunk 3 above)
Comment (francescogior) [reply src/index.js]: ¯\_(ツ)_/¯

File: src/index.js (hunk 5 above)
Comment (francescogior) [reply src/index.js:134]: No wait, I think you misread it as `stripUndef(readyState)`.

Comment (FrancescoCioria) [reply src/index.js:134]: oh right, basically you're adding `readyState` only if it's not `undefined`. 👍

File: src/index.js (hunk 3 above)
Comment (francescogior) [reply src/index.js]: Oh, btw I just discovered that there was a bug here. I needed `omitBy` from `lodash/fp/omitBy`

**Review by FrancescoCioria — APPROVED** [review]

File: src/index.js [hunk 6]
```diff
@@ -31,12 +36,16 @@ const PublicDecoratorConfig = DecoratorConfig.extend({
...
 const reduceQueryPropsReturn = queries => t.interface({
   accumulator: t.Any,
-  props: t.interface(
-    queries.reduce((ac, k) => ({ ...ac, [k]: t.Any }), {}),
-    { strict: true }
-  )
+  props: t.interface({
+    ...queries.reduce((ac, k) => ({ ...ac, [k]: t.Any }), {}),
+    readyState: t.interface(queries.reduce((ac, k) => ({ ...ac, [k]: ReadyState }), {}), { strict: true })
+  }, { strict: true, name: 'ReadyStates' })
```
Comment (giogonzo) [inline src/index.js]: this is not just `ReadyStates` right? It should be named something like `QueryProps` maybe?

File: src/index.js [hunk 7]
```diff
@@ -109,12 +117,22 @@ const decorator = ({ declareQueries, declareCommands, declareConnect }) => (Comp
...
+
+    getLocals(props) {
+      const { readyState } = props;
+      if (isReady(props)) {
+        return { ...stripUndef({ readyState }), ...getLocals(props) };
+      } else {
+        return stripUndef({ readyState });
+      }
+    }
+
```
Comment (giogonzo) [inline src/index.js:137]: what happens when no `queries` are defined on the container? Probably it works just fine (no `readyState` passed, `isReady` always return `true`, we always end up in the `then` branch), but maybe everything's a bit too implicit?

**Review by giogonzo — COMMENTED** [review]
same here, just two inline comments

File: src/index.js (hunk 6 above)
Comment (francescogior) [reply src/index.js]: yes, wrong line! `ReadyStates` is the one above, this one is `QueriesProps` or something

File: src/index.js (hunk 7 above)
Comment (francescogior) [reply src/index.js:137]: Yes, because `every({}) === true`. But ok to make it more explicit

**Pull request merged**